README.md
# Sensirion I2C SHT3X embedded Library

This document explains how to set up a sensor of the SHT3X family to run on an embedded device using the I²C interface.

<img src="images/SHT3x.png" width="300px">

Click [here](https://sensirion.com/products/catalog/SHT30-DIS-B) to learn more about the Sensirion SHT3X sensor family.


Not all sensors of this driver family support all measurements.
In case a measurement is not supported by all sensors, the products that
support it are listed in the API description.



## Supported sensor types

| Sensor name   | I²C Addresses  |
| ------------- | -------------- |
|[SHT30A](https://sensirion.com/products/catalog/SHT30A-DIS-B)| **0x44**, 0x45|
|[SHT30](https://sensirion.com/products/catalog/SHT30-DIS-B)| **0x44**, 0x45|
|[SHT31A](https://sensirion.com/products/catalog/SHT31A-DIS-B)| **0x44**, 0x45|
|[SHT31](https://sensirion.com/products/catalog/SHT31-DIS-B)| **0x44**, 0x45|
|[SHT33](https://sensirion.com/products/catalog/SHT33-DIS)| **0x44**, 0x45|
|[SHT35A](https://sensirion.com/products/catalog/SHT35A-DIS-B)| **0x44**, 0x45|
|[SHT35](https://sensirion.com/products/catalog/SHT35-DIS-B)| **0x44**, 0x45|
|[SHT85](https://sensirion.com/sht85/)| **0x44**|

The following instructions and examples use a *SHT30*.



## Setup Guide

### Connecting the Sensor

Your sensor has 4 different signals that need to be connected to your board: SDA, GND, SCL, VDD.
Use the following pins to connect your SHT3X:

<img src="images/SHT3x_pinout.png" width="300px">

| *Pin* | *Cable Color* | *Name* | *Description*  | *Comments* |
|-------|---------------|:------:|----------------|------------|
| 1 | green | SDA | I2C: Serial data input / output | 
| 2 | black | GND | Ground | 
| 3 | yellow | SCL | I2C: Serial clock input | 
| 4 | red | VDD | Supply Voltage | 2.15V to 5.5V



The recommended voltage is 3.3V.

### Configure the code

In order to use the provided code we need to adjust two files according to your platform.

### Edit `sensirion_i2c_hal.c`

This file contains the implementation of the sensor communication, which
depends on your hardware platform. We provide function stubs for your
hardware's own implementation.
Sample implementations are available for some platforms:
[`sample-implementations`](sample-implementations). For Linux based platforms
like Raspberry Pi you can just replace the unimplemented HAL template with the
implementation in `sample-implementations/linux_user_space/`:

```
cp sample-implementations/linux_user_space/sensirion_i2c_hal.c ./
```

### Edit `sensirion_config.h`

Skip this part for Linux based platforms since everything is already setup for
this case.

Otherwise you need to check if the libraries `<stdint.h>` and `<stdlib.h>` are
provided by your toolchain, compiler or system. If you have no idea on how to
do that you can skip this step for now and come back when you get errors
related to these names when compiling the driver.
The features we use from those libraries are type definitions for integer sizes
from `<stdint.h>` and `NULL` from `<stdlib.h>`. If they are not available you
need to specify the following integer types yourself:

* `int64_t` = signed 64bit integer
* `uint64_t` = unsigned 64bit integer
* `int32_t` = signed 32bit integer
* `uint32_t` = unsigned 32bit integer
* `int16_t` = signed 16bit integer
* `uint16_t` = unsigned 16bit integer
* `int8_t` = signed 8bit integer
* `uint8_t` = unsigned 8bit integer

In addition to that you will need to specify `NULL`. For both we have a
detailed template where you just need to fill in your system specific values.

## Choose the i2c address to use with your product

The provided example is working with a SHT30, I²C address 0x44.
In order to use the code with another product or I²C address you need to change it in the call sht3x_init(ADDRESS) in
`sht3x_i2c_example_usage.c`. The list of supported I²C-addresses is found in the header 
`sht3x_i2c.h`.


Now we are ready to compile and run the example usage for your sensor.

## Compile and Run

Pass the source `.c` and header `.h` files in this folder into your C compiler
and run the resulting binary. This step may vary, depending on your platform.
Here we demonstrate the procedure for Linux based platforms:

1. Open up a terminal.
2. Navigate to the directory where this README is located.
3. Navigate to the subdirectory example-usage.
4. Run `make` (this compiles the example code into one executable binary).
5. Run the compiled executable with `./sht3x_i2c_example_usage`
6. Now you should see the first measurement values appear in your terminal. As
   a next step you can adjust the example usage file or write your own main
   function to use the sensor.

## Compile and Run Tests

The testframekwork used is CppUTest. Pass the source `.cpp`, `.c`  and header `.h`
files from the tests and top level folder into your CPP compiler and run the
resulting binary. This step may vary, depending on your platform.
Here we demonstrate the procedure for Linux based platforms:

1. Open up a terminal.
2. Install CppUTest framework `apt install cpputest`.
3. Navigate to the directory `tests`.
4. Run `make` (this compiles the test code into one executable binary).
5. Run the compiled executable with `./sht3x_test_hw_i2c`.
6. Now you should see the test output on your console.

# Background

## Files

### sensirion\_i2c.[ch]

In these files you can find the implementation of the I2C protocol used by
Sensirion sensors. The functions in these files are used by the embedded driver
to build the correct frame out of data to be sent to the sensor or receive a
frame of data from the sensor and convert it back to data readable by your
machine. The functions in here calculate and check CRCs, reorder bytes for
different byte orders and build the correct formatted frame for your sensor.

### sensirion\_i2c\_hal.[ch]

These files contain the implementation of the hardware abstraction layer used
by Sensirion's I2C embedded drivers. This part of the code is specific to the
underlying hardware platform. This is an unimplemented template for the user to
implement. In the `sample-implementations/` folder we provide implementations
for the most common platforms.

### sensirion\_config.h

In this file we keep all the included libraries for our drivers and global
defines. Next to `sensirion_i2c_hal.c` *it's the only file you should need to
edit to get your driver working.*

### sensirion\_common.[ch]

In these files you can find some helper functions used by Sensirion's embedded
drivers. It mostly contains byte order conversions for different variable
types. These functions are also used by the UART embedded drivers therefore
they are kept in their own file.

## Contributing

**Contributions are welcome!**

We develop and test this driver using our company internal tools (version
control, continuous integration, code review etc.) and automatically
synchronize the master branch with GitHub. But this doesn't mean that we don't
respond to issues or don't accept pull requests on GitHub. In fact, you're very
welcome to open issues or create pull requests :)

This Sensirion library uses
[`clang-format`](https://releases.llvm.org/download.html) to standardize the
formatting of all our `.c` and `.h` files. Make sure your contributions are
formatted accordingly:

The `-i` flag will apply the format changes to the files listed.

```bash
clang-format -i *.c *.h
```

Note that differences from this formatting will result in a failed build until
they are fixed.


# License

See [LICENSE](LICENSE).

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]

## [2.0.0] - 2023-11-13

### CHANGED

- Remove usage of float.
  sht3x_measure_single_shot and sht3x_blocking_read_measurement now return a int32_t,
  conversion from ticks to physical value uses an approximation and the returned
  temperature is in milli degrees celsius and humidity in milli %.

## [1.0.0] - 2023-10-27

Initial release



example-usage/sht3x_i2c_example_usage.c
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * Generator:     sensirion-driver-generator 0.33.0
 * Product:       sht3x
 * Model-Version: 2.0.0
 */
/*
 * Copyright (c) 2023, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include "sensirion_common.h"
#include "sensirion_i2c_hal.h"
#include "sht3x_i2c.h"
#include <stdio.h>  // printf

#define sensirion_hal_sleep_us sensirion_i2c_hal_sleep_usec

int main(void) {
    int16_t error = NO_ERROR;
    sensirion_i2c_hal_init();
    sht3x_init(SHT30_I2C_ADDR_44);

    sht3x_stop_measurement();
    sensirion_hal_sleep_us(1000);
    sht3x_soft_reset();
    sensirion_hal_sleep_us(100000);
    uint16_t a_status_register = 0u;
    error = sht3x_read_status_register(&a_status_register);
    if (error != NO_ERROR) {
        printf("error executing read_status_register(): %i\n", error);
        return error;
    }
    printf("a_status_register: %02x\n", a_status_register);
    error = sht3x_start_periodic_measurement(REPEATABILITY_MEDIUM,
                                             MPS_ONE_PER_SECOND);
    if (error != NO_ERROR) {
        printf("error executing start_periodic_measurement(): %i\n", error);
        return error;
    }
    int32_t a_temperature = 0.0;
    int32_t a_humidity = 0.0;
    uint16_t repetition = 0;
    for (repetition = 0; repetition < 50; repetition++) {
        error = sht3x_blocking_read_measurement(&a_temperature, &a_humidity);
        if (error != NO_ERROR) {
            printf("error executing blocking_read_measurement(): %i\n", error);
            continue;
        }
        printf("a_temperature [milli degC]: %i ", a_temperature);
        printf("a_humidity [milli RH]: %i\n", a_humidity);
    }

    error = sht3x_stop_measurement();
    if (error != NO_ERROR) {
        return error;
    }

    return NO_ERROR;
}


example-usage-single-shot/sht3x_i2c_example_usage_single_shot.c
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * Generator:     sensirion-driver-generator 0.33.0
 * Product:       sht3x
 * Model-Version: 2.0.0
 */
/*
 * Copyright (c) 2023, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include "sensirion_common.h"
#include "sensirion_i2c_hal.h"
#include "sht3x_i2c.h"
#include <stdio.h>  // printf

#define sensirion_hal_sleep_us sensirion_i2c_hal_sleep_usec

int main(void) {
    int16_t error = NO_ERROR;
    sensirion_i2c_hal_init();
    sht3x_init(SHT30_I2C_ADDR_44);

    sht3x_stop_measurement();
    sensirion_hal_sleep_us(1000);
    sht3x_soft_reset();
    sensirion_hal_sleep_us(100000);
    uint16_t a_status_register = 0u;
    error = sht3x_read_status_register(&a_status_register);
    if (error != NO_ERROR) {
        printf("error executing read_status_register(): %i\n", error);
        return error;
    }
    printf("a_status_register: %02x\n", a_status_register);
    int32_t a_temperature = 0;
    int32_t a_humidity = 0;
    uint16_t repetition = 0;
    for (repetition = 0; repetition < 50; repetition++) {
        error = sht3x_measure_single_shot(REPEATABILITY_MEDIUM, false,
                                          &a_temperature, &a_humidity);
        if (error != NO_ERROR) {
            printf("error executing measure_single_shot(): %i\n", error);
            continue;
        }
        printf("a_temperature  [milli degC]: %i ", a_temperature);
        printf("a_humidity  [milli RH]: %i\n", a_humidity);
    }

    return NO_ERROR;
}


sht3x_i2c.h
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * Generator:     sensirion-driver-generator 0.33.0
 * Product:       sht3x
 * Model-Version: 2.0.0
 */
/*
 * Copyright (c) 2023, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SHT3X_I2C_H
#define SHT3X_I2C_H

#ifdef __cplusplus
extern "C" {
#endif

#include "sensirion_config.h"
#define SHT30A_I2C_ADDR_44 0x44
#define SHT30A_I2C_ADDR_45 0x45
#define SHT30_I2C_ADDR_44 0x44
#define SHT30_I2C_ADDR_45 0x45
#define SHT31A_I2C_ADDR_44 0x44
#define SHT31A_I2C_ADDR_45 0x45
#define SHT31_I2C_ADDR_44 0x44
#define SHT31_I2C_ADDR_45 0x45
#define SHT33_I2C_ADDR_44 0x44
#define SHT33_I2C_ADDR_45 0x45
#define SHT35A_I2C_ADDR_44 0x44
#define SHT35A_I2C_ADDR_45 0x45
#define SHT35_I2C_ADDR_44 0x44
#define SHT35_I2C_ADDR_45 0x45
#define SHT85_I2C_ADDR_44 0x44

typedef enum {
    MEASURE_SINGLE_SHOT_HIGH_REPEATABILITY_CMD_ID = 0x2400,
    MEASURE_SINGLE_SHOT_HIGH_REPEATABILITY_CLOCK_STRETCHING_CMD_ID = 0x2c06,
    MEASURE_SINGLE_SHOT_MEDIUM_REPEATABILITY_CMD_ID = 0x240b,
    MEASURE_SINGLE_SHOT_MEDIUM_REPEATABILITY_CLOCK_STRETCHING_CMD_ID = 0x2c0d,
    MEASURE_SINGLE_SHOT_LOW_REPEATABILITY_CMD_ID = 0x2416,
    MEASURE_SINGLE_SHOT_LOW_REPEATABILITY_CLOCK_STRETCHING_CMD_ID = 0x2c10,
    START_MEASUREMENT_0_5_MPS_HIGH_REPEATABILITY_CMD_ID = 0x2032,
    START_MEASUREMENT_0_5_MPS_MEDIUM_REPEATABILITY_CMD_ID = 0x2024,
    START_MEASUREMENT_0_5_MPS_LOW_REPEATABILITY_CMD_ID = 0x202f,
    START_MEASUREMENT_1_MPS_HIGH_REPEATABILITY_CMD_ID = 0x2130,
    START_MEASUREMENT_1_MPS_MEDIUM_REPEATABILITY_CMD_ID = 0x2126,
    START_MEASUREMENT_1_MPS_LOW_REPEATABILITY_CMD_ID = 0x212d,
    START_MEASUREMENT_2_MPS_HIGH_REPEATABILITY_CMD_ID = 0x2236,
    START_MEASUREMENT_2_MPS_MEDIUM_REPEATABILITY_CMD_ID = 0x2220,
    START_MEASUREMENT_2_MPS_LOW_REPEATABILITY_CMD_ID = 0x222b,
    START_MEASUREMENT_4_MPS_HIGH_REPEATABILITY_CMD_ID = 0x2334,
    START_MEASUREMENT_4_MPS_MEDIUM_REPEATABILITY_CMD_ID = 0x2322,
    START_MEASUREMENT_4_MPS_LOW_REPEATABILITY_CMD_ID = 0x2329,
    START_MEASUREMENT_10_MPS_HIGH_REPEATABILITY_CMD_ID = 0x2737,
    START_MEASUREMENT_10_MPS_MEDIUM_REPEATABILITY_CMD_ID = 0x2721,
    START_MEASUREMENT_10_MPS_LOW_REPEATABILITY_CMD_ID = 0x273a,
    START_ART_MEASUREMENT_CMD_ID = 0x2b32,
    READ_MEASUREMENT_CMD_ID = 0xe000,
    STOP_MEASUREMENT_CMD_ID = 0x3093,
    ENABLE_HEATER_CMD_ID = 0x306d,
    DISABLE_HEATER_CMD_ID = 0x3066,
    READ_STATUS_REGISTER_CMD_ID = 0xf32d,
    CLEAR_STATUS_REGISTER_CMD_ID = 0x3041,
    SOFT_RESET_CMD_ID = 0x30a2,
} cmd_id_t;

typedef enum {
    REPEATABILITY_LOW = 0,
    REPEATABILITY_MEDIUM = 1,
    REPEATABILITY_HIGH = 2,
} repeatability;

typedef enum {
    MPS_EVERY_TWO_SECONDS = 0,
    MPS_ONE_PER_SECOND = 1,
    MPS_TWO_PER_SECOND = 2,
    MPS_FOUR_PER_SECOND = 4,
    MPS_TEN_PER_SECOND = 10,
} mps;

/**
 * @brief Initialize i2c address of driver
 *
 * @param[in] i2c_address Used i2c address
 *
 */
void sht3x_init(uint8_t i2c_address);

/**
 * @brief Convert temperature ticks to physical value (approximation)
 *
 * @param[in] temperature_ticks
 *
 * @return temperature in milli degrees celsius
 */
int32_t signal_temperature(uint16_t temperature_ticks);

/**
 * @brief Convert humidity ticks to physical value (approximation)
 *
 * @param[in] humidity_ticks
 *
 * @return relative humidity in milli percent RH
 */
int32_t signal_humidity(uint16_t humidity_ticks);

/**
 * @brief Single shot measurement with the specified properties
 *
 * @param[in] measurement_repeatability The repeatability of the periodic
 * measurement
 * @param[in] is_clock_stretching Toggle clock stretching
 * @param[out] a_temperature Measured temperature in milli degree celsius
 * @param[out] a_humidity Measured humidity in milli percent RH
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_measure_single_shot(repeatability measurement_repeatability,
                                  bool is_clock_stretching,
                                  int32_t* a_temperature, int32_t* a_humidity);

/**
 * @brief sht3x_start_periodic_measurement
 *
 * Start the periodic measurement measurement mode.
 *
 * This is a convenience method that selects the correct measurement command
 * based on the provided arguments.
 *
 * @param[in] measurement_repeatability The repeatability of the periodic
 * measurement
 * @param[in] messages_per_second The messages per second of the periodic
 * measurement
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t
sht3x_start_periodic_measurement(repeatability measurement_repeatability,
                                 mps messages_per_second);

/**
 * @brief sht3x_blocking_read_measurement
 *
 * This is a convenience method that combines polling the data ready flag and
 * reading out the data. As the minimal measurement interval is 2s and we sleep
 * for 100ms we iterate at most 200 times. Note that this is blocking the system
 * for a considerable amount of time!
 *
 * @param[out] a_temperature Measured temperature in milli degree celsius
 * @param[out] a_humidity Measured humidity in milli percent RH
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_blocking_read_measurement(int32_t* a_temperature,
                                        int32_t* a_humidity);

/**
 * @brief Read the contents of the status register
 *
 * @param[out] a_status_register
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_read_status_register(uint16_t* a_status_register);

/**
 * @brief sht3x_measure_single_shot_high_repeatability
 *
 * Single shot measurement with high repeatability
 *
 * @param[out] temperature_ticks Temperature ticks. Convert to degrees celsius
 * by -45 + 175 * value / 65535
 * @param[out] humidity_ticks Humidity ticks. Convert to relative humidity by
 * 100 * value / 65535
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t
sht3x_measure_single_shot_high_repeatability(uint16_t* temperature_ticks,
                                             uint16_t* humidity_ticks);

/**
 * @brief sht3x_measure_single_shot_high_repeatability_clock_stretching
 *
 * Single shot measurement with high repeatability and clock stretching enabled
 *
 * @param[out] temperature_ticks Temperature ticks. Convert to degrees celsius
 * by -45 + 175 * value / 65535
 * @param[out] humidity_ticks Humidity ticks. Convert to relative humidity by
 * 100 * value / 65535
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_measure_single_shot_high_repeatability_clock_stretching(
    uint16_t* temperature_ticks, uint16_t* humidity_ticks);

/**
 * @brief sht3x_measure_single_shot_medium_repeatability
 *
 * Single shot measurement with medium repeatability
 *
 * @param[out] temperature_ticks Temperature ticks. Convert to degrees celsius
 * by -45 + 175 * value / 65535
 * @param[out] humidity_ticks Humidity ticks. Convert to relative humidity by
 * 100 * value / 65535
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t
sht3x_measure_single_shot_medium_repeatability(uint16_t* temperature_ticks,
                                               uint16_t* humidity_ticks);

/**
 * @brief sht3x_measure_single_shot_medium_repeatability_clock_stretching
 *
 * Single shot measurement with medium repeatability and clock stretching
 * enabled
 *
 * @param[out] temperature_ticks Temperature ticks. Convert to degrees celsius
 * by -45 + 175 * value / 65535
 * @param[out] humidity_ticks Humidity ticks. Convert to relative humidity by
 * 100 * value / 65535
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_measure_single_shot_medium_repeatability_clock_stretching(
    uint16_t* temperature_ticks, uint16_t* humidity_ticks);

/**
 * @brief sht3x_measure_single_shot_low_repeatability
 *
 * Single shot measurement with low repeatability
 *
 * @param[out] temperature_ticks Temperature ticks. Convert to degrees celsius
 * by -45 + 175 * value / 65535
 * @param[out] humidity_ticks Humidity ticks. Convert to relative humidity by
 * 100 * value / 65535
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_measure_single_shot_low_repeatability(uint16_t* temperature_ticks,
                                                    uint16_t* humidity_ticks);

/**
 * @brief sht3x_measure_single_shot_low_repeatability_clock_stretching
 *
 * Single shot measurement with low repeatability and clock stretching enabled
 *
 * @param[out] temperature_ticks Temperature ticks. Convert to degrees celsius
 * by -45 + 175 * value / 65535
 * @param[out] humidity_ticks Humidity ticks. Convert to relative humidity by
 * 100 * value / 65535
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_measure_single_shot_low_repeatability_clock_stretching(
    uint16_t* temperature_ticks, uint16_t* humidity_ticks);

/**
 * @brief sht3x_start_measurement_0_5_mps_high_repeatability
 *
 * Start periodic measurement mode with 0.5 mps and high repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_0_5_mps_high_repeatability();

/**
 * @brief sht3x_start_measurement_0_5_mps_medium_repeatability
 *
 * Start periodic measurement mode with 0.5 mps and medium repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_0_5_mps_medium_repeatability();

/**
 * @brief sht3x_start_measurement_0_5_mps_low_repeatability
 *
 * Start periodic measurement mode with 0.5 mps and low repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_0_5_mps_low_repeatability();

/**
 * @brief sht3x_start_measurement_1_mps_high_repeatability
 *
 * Start periodic measurement mode with 1 mps and high repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_1_mps_high_repeatability();

/**
 * @brief sht3x_start_measurement_1_mps_medium_repeatability
 *
 * Start periodic measurement mode with 1 mps and medium repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_1_mps_medium_repeatability();

/**
 * @brief sht3x_start_measurement_1_mps_low_repeatability
 *
 * Start periodic measurement mode with 1 mps and low repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_1_mps_low_repeatability();

/**
 * @brief sht3x_start_measurement_2_mps_high_repeatability
 *
 * Start periodic measurement mode with 2 mps and high repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_2_mps_high_repeatability();

/**
 * @brief sht3x_start_measurement_2_mps_medium_repeatability
 *
 * Start periodic measurement mode with 2 mps and medium repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_2_mps_medium_repeatability();

/**
 * @brief sht3x_start_measurement_2_mps_low_repeatability
 *
 * Start periodic measurement mode with 2 mps and low repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_2_mps_low_repeatability();

/**
 * @brief sht3x_start_measurement_4_mps_high_repeatability
 *
 * Start periodic measurement mode with 4 mps and high repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_4_mps_high_repeatability();

/**
 * @brief sht3x_start_measurement_4_mps_medium_repeatability
 *
 * Start periodic measurement mode with 4 mps and medium repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_4_mps_medium_repeatability();

/**
 * @brief sht3x_start_measurement_4_mps_low_repeatability
 *
 * Start periodic measurement mode with 4 mps and low repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_4_mps_low_repeatability();

/**
 * @brief sht3x_start_measurement_10_mps_high_repeatability
 *
 * Start periodic measurement mode with 10 mps and high repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_10_mps_high_repeatability();

/**
 * @brief sht3x_start_measurement_10_mps_medium_repeatability
 *
 * Start periodic measurement mode with 10 mps and medium repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_10_mps_medium_repeatability();

/**
 * @brief sht3x_start_measurement_10_mps_low_repeatability
 *
 * Start periodic measurement mode with 10 mps and low repeatability.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_measurement_10_mps_low_repeatability();

/**
 * @brief sht3x_start_art_measurement
 *
 * Start ART (accelerated response time) measurement
 *
 * @note After issuing the ART command the sensor will start acquiring data with
 * a frequency of 4Hz.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_start_art_measurement();

/**
 * @brief sht3x_read_measurement
 *
 * Read out data after a "start measurement" or "start art measurement" command
 * has been issued.
 *
 * @param[out] temperature_ticks Temperature ticks. Convert to degrees celsius
 * by -45 + 175 * value / 65535
 * @param[out] humidity_ticks Humidity ticks. Convert to relative humidity by
 * 100 * value / 65535
 *
 * @note After the read out command fetch data has been issued, the data memory
 * is cleared, i.e. no measurement data is present.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_read_measurement(uint16_t* temperature_ticks,
                               uint16_t* humidity_ticks);

/**
 * @brief sht3x_stop_measurement
 *
 * Stop the periodic measurement mode.
 *
 * @note Upon reception of this command the sensor will abort the ongoing
 * measurement and enter the single shot mode.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_stop_measurement();

/**
 * @brief sht3x_enable_heater
 *
 * Enable the heater
 *
 * @note The SHT3x is equipped with an internal heater, which is meant for
 * plausibility checking only. The temperature increase achieved by the heater
 * depends on various parameters and lies in the range of a few degrees
 * centigrade.
 *
 * After a reset the heater is disabled (default condition).
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_enable_heater();

/**
 * @brief sht3x_disable_heater
 *
 * Disable the heater
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_disable_heater();

/**
 * @brief ll_sht3x_read_status_register
 *
 * Read out the status register
 *
 * @param[out] status_register The contents of the status register
 *
 * @note The status register contains information on the operational status of
 * the heater, the alert mode and on the execution status of the last command
 * and the last write sequence.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t ll_sht3x_read_status_register(uint16_t* status_register);

/**
 * @brief sht3x_clear_status_register
 *
 * Clear (set to zero) all flags (Bit 4) in the status register.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_clear_status_register();

/**
 * @brief sht3x_soft_reset
 *
 * Perform a soft reset.
 *
 * @note A system reset of the SHT3x can be achieved in multiple ways: - Soft
 * reset: use this command - I2C general call: all devices on the I2C bus are
 * reset by sending the command 0x06 to the I2C address 0x00 - Reset pin: send a
 * pulse to the dedicated nReset pin. The nReset pin has to be pulled low for a
 * minimum of 1 µs to generate a reset of the sensor. - Hard reset: Power down
 * (incl. pulling SDA, SCL and ADDR low)
 *
 * During the reset procedure the sensor will not process commands.
 *
 * @return error_code 0 on success, an error code otherwise.
 */
int16_t sht3x_soft_reset();

#ifdef __cplusplus
}
#endif
#endif  // SHT3X_I2C_H
